src/main/java/org/jenkinsci/remoting/util/AnonymousClassWarnings.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2018 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.remoting.util;
+
+import hudson.remoting.Channel;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.security.CodeSource;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * Issues warnings about attempts to (de-)serialize anonymous, local, or synthetic classes.
+ * @see <a href="JENKINS-49994">https://issues.jenkins-ci.org/browse/JENKINS-49994</a>
+ */
+public class AnonymousClassWarnings {
+
+    private static final Logger LOGGER = Logger.getLogger(AnonymousClassWarnings.class.getName());    
+    private static final Map<Class<?>, Boolean> checked = new WeakHashMap<>();
+
+    /**
+     * Checks a class which is being either serialized or deserialized.
+     * A warning will only be printed once per class per JVM session.
+     */
+    public static void check(@Nonnull Class<?> clazz) {
+        synchronized (checked) {
+            if (checked.containsKey(clazz)) {
+                return; // fast path
+            }
+        }
+        Channel channel = Channel.current();
+        if (channel == null) {
+            doCheck(clazz);
+        } else {
+            // May not call methods like Class#isAnonymousClass synchronously, since these can in turn trigger remote class loading.
+            try {
+                channel.executor.submit((Runnable) () -> doCheck(clazz));
```

Comment:
Having this `ExecutorService` be passed to `check` would be cleaner, but I did not see a reasonable way to do that—a ton of other APIs would have to be modified to thread this parameter in. The use of `Channel.current()` suffices for our purposes.